modules/artifact-activation/src/main/java/org/mule/runtime/module/artifact/activation/internal/extension/MuleExtensionModelLoaderManager.java

```diff
@@ -0,0 +1,143 @@
+/*
+ * Copyright (c) MuleSoft, Inc.  All rights reserved.  http://www.mulesoft.com
+ * The software in this package is published under the terms of the CPAL v1.0
+ * license, a copy of which has been included with this distribution in the
+ * LICENSE.txt file.
+ */
+
+package org.mule.runtime.module.artifact.activation.internal.extension;
+
+import org.mule.runtime.api.exception.MuleException;
+import org.mule.runtime.api.exception.MuleRuntimeException;
+import org.mule.runtime.api.lifecycle.Startable;
+import org.mule.runtime.api.lifecycle.Stoppable;
+import org.mule.runtime.extension.api.loader.ExtensionModelLoader;
+import org.mule.runtime.extension.api.loader.ExtensionModelLoaderProvider;
+import org.mule.runtime.module.artifact.activation.api.extension.ExtensionModelLoaderRepository;
+import org.mule.runtime.module.artifact.api.plugin.LoaderDescriber;
+import org.slf4j.Logger;
+
+import java.util.Collection;
+import java.util.Map;
+import java.util.Optional;
+import java.util.function.Supplier;
+
+import static com.google.common.collect.Maps.newHashMap;
+import static java.lang.String.format;
+import static java.lang.System.lineSeparator;
+import static java.util.Objects.requireNonNull;
+import static java.util.Optional.empty;
+import static java.util.Optional.of;
+import static java.util.stream.Collectors.*;
+import static org.apache.commons.lang3.StringUtils.isNotBlank;
+import static org.mule.runtime.api.i18n.I18nMessageFactory.createStaticMessage;
+import static org.mule.runtime.core.api.util.boot.ExtensionLoaderUtils.lookupExtensionModelLoaders;
+import static org.slf4j.LoggerFactory.getLogger;
+
+/**
+ * Implementation of {@link ExtensionModelLoaderRepository} that uses SPI to look for the {@link ExtensionModelLoader} available
+ * from the container.
+ *
+ * @since 4.5
+ */
+public class MuleExtensionModelLoaderManager implements ExtensionModelLoaderRepository, Startable, Stoppable {
+
+  private static final Logger LOGGER = getLogger(MuleExtensionModelLoaderManager.class);
+
+  private final Map<String, ExtensionModelLoader> extensionModelLoaders = newHashMap();
+
+  private Supplier<Collection<ExtensionModelLoader>> extModelLoadersLookup;
+
+  /**
+   * Creates an instance of the manager.
+   *
+   * @param containerClassLoader {@link ClassLoader} from the container.
+   */
+  public MuleExtensionModelLoaderManager(ClassLoader containerClassLoader) {
+    requireNonNull(containerClassLoader, "containerClassLoader cannot be null");
+    this.extModelLoadersLookup = () -> lookupLoadersFromSpi(containerClassLoader);
```

Comment:
why a supplier if not lazy and if there's no way you end up not eventually using these loaders?